### DISCUSSION TITLE
is extending the FastAPI class for adding properties and other stuff okay

### DISCUSSION THREAD

**Discussion by iyad-f** [fastapi/fastapi#12447] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class Application(FastAPI):
    """The main application extends FastAPI to provide more functionality"""

    def __init__(self, *, pool: Pool[Any | Record], session: ClientSession) -> None:
        self._pool: Pool[Any | Record] = pool
        self._session: ClientSession = session

        super().__init__(
            title="codex", openapi_url=f'/api/v{CONFIG["API"]["version"]}/openapi.json', lifespan=self.lifespan_handler
        )

        self.include_router(v1.ROUTER)

    @property
    def pool(self) -> Pool[Any | Record]:
        """:class:`asyncpg.Pool` The database connection pool to make database queries"""
        return self._pool

    @property
    def session(self) -> ClientSession:
        """:class:`aiohttp.ClientSession` The client session to make http requests"""
        return self._session

    # NOTE: the order of those decorators matters
    @staticmethod
    @asynccontextmanager
    async def lifespan_handler(app: Application) -> AsyncIterator[None]:
        """An async context manager to handle the lifespan of the app"""
        await app._setup_db()

        yield

        await app.session.close()
        await app.pool.close()

    async def _setup_db(self) -> None:
        """Sets up the postgresql database"""
        async with self.pool.acquire() as conn:
            for path in pathlib.Path("sql").iterdir():
                with open(path) as file:
                    await conn.execute(file.read())
```

### Description

i have currently implemented the given code but i am wondering that i shouldn't do this because i saw such object oriented approach no where should just make the app instance normally and then set app.pool and app.session after instantiating it as just app = FastAPI() but or store those in the app.state but then again i lose those type hinting perks and also the fact that i loose the ability to make those properties and document them as i have in my original approach, so what should i do, is doing this okay ?

### Operating System

Windows

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.7

**Comment by Kfir-G** [comment, marked as answer]
@iyad-f Absolutely, extending the FastAPI class is a great idea. It keeps your code organized, lets you use type hints, and makes resource management super clean. 

Sure, you could use `app.state` or composition, but you’d lose some of those benefits. So if it works for you, go for it! Your approach is totally valid.

**Comment by iyad-f** [comment]
thankyou @Kfir-G, i just wanted to confirm this because the way fastapi is build and most of the docs out there just suggest the decorator approach so in my case if i have to use a decorator lets say app.exception_handler in my __init__ only so i would have to do self.exception_handler(HTTPException)(self._http_exc_handler), so basically would just go against the normal convention of fastapi does things but ya this way i do get the type checking benefits so i think i am gonna go with it.
